### PR TITLE
TurboQuant encoding for Vectors

### PR DESCRIPTION
**Pull request by connortsui20**
Continuation of https://github.com/vortex-data/vortex/pull/7167, authored by @lwwmanning 

## Summary

Lossy quantization for vector data (e.g., embeddings) based on TurboQuant (https://arxiv.org/abs/2504.19874). Implements the MSE-only variant (Stage 1 of RFC 0033) at 1-8 bits per coordinate (0 for empty arrays), defaulting to 8-bit near-lossless compression (but still a small amount because we use SRHT instead of random orthogonal rotation matrix, something about not satisfying the Haar assumption?).

Key components:
- TurboQuant array encoding with 4 slots: quantized codes, norms, centroids, and rotation signs. Note that we should probably abstract the codes and centroids as a dictionary encoded thing so we don't have to duplicate pushdown rules, and we might want to make a matrix multiplication `ScalarFn` expression for the rotations.
- Structured Random Hadamard Transform (SRHT) for `O(d log d)` rotation, fully self-contained with no external linear algebra library. This is what claude came up with, but we can see in testing that while this is practical and more efficient, we lose some of the assumptions that a Haar-random orthogonal matrix gives us. I think this is something we can play around with because it's abstracted into a discrete step of the algorithm.
- Max-Lloyd centroid computation on the Beta distribution for the given dimension.
- Approximate cosine similarity and dot product computed directly on quantized arrays without full decompression.
- Pluggable `TurboQuantScheme` for the cascading compressor.
- Minimum dimension of 128 (`TurboQuant::MIN_DIMENSION`) for SRHT quality guarantees.
- Default 8-bit encoding (MSE ~4e-5, exact 4x compression on `f32`).
- Adds `vortex_tensor::initialize()` for session registration of tensor types, encodings, and scalar functions.

## API Changes

- Adds `TurboQuant` encoding in `vortex-tensor` with `turboquant_encode()` and `TurboQuantConfig`, and new types `TurboQuantData` and `TurboQuantArray`.
- Adds `TurboQuantScheme` for compressor integration.
- Adds `TurboQuant::MIN_DIMENSION` (128) constant.
- Adds `float_from_f32<T: Float + FromPrimitive>` shared helper for infallible f32-to-float conversion.

## Testing (claude-generated)

- Roundtrip tests across bit widths (1-8) and dimensions (128, 256, 768, 1024).
- MSE quality bounds verified against the theoretical bound from Theorem 1.
- Edge cases: empty arrays, single-row arrays, all-zero vectors, dimension rejection below 128.
- Float type coverage: f16, f32, f64 input encoding and roundtrip.
- Nullable vector support: validity propagation through encode, decode, slice, take, L2 norm readthrough.
- Quantized-domain cosine similarity and dot product accuracy tests.
- Serde roundtrip (serialize/deserialize).
- Compute pushdown tests: slice, take, scalar_at.
- Compression ratio estimates for typical embedding dimensions.
- Centroid correctness: count, sorted, symmetric, within bounds, caching, boundary rejection.
- SRHT rotation: determinism, roundtrip, norm preservation, sign export/import roundtrip.

### CODE REVIEW

**Comment by connortsui20** [comment]
I would say this is ready for review now, but only with respect to the structure. I have yet to go through the implementation and make sure things make sense, but I have made it so the structure makes sense and we correctly handle the different floating point type inputs as well as null vectors.

I think it would be good to get a review now, and maybe we should just merge this and iterate later.

**Review by gatesn — APPROVED** [review]
Let's do it

**Comment by connortsui20** [comment]
I'll rebase and fix the errors tomorrow

File: vortex-tensor/src/encodings/turboquant/compress.rs [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    /// Bits per coordinate (1-8).
+    pub bit_width: u8,
+    /// Optional seed for the rotation matrix. If None, the default seed is used.
+    pub seed: Option<u64>,
+}
+
+impl Default for TurboQuantConfig {
+    fn default() -> Self {
+        Self {
+            bit_width: 4,
```
Comment (lwwmanning) [inline vortex-tensor/src/encodings/turboquant/compress.rs]: RFC 0033 Stage 1 specifies "5-bit MSE-only (32 centroids)" as the scheme default. This currently defaults to 4-bit (16 centroids).

From the RFC table:

> | Scheme default | 5-bit QJL (4-bit MSE + 1-bit QJL) | **5-bit MSE-only** (32 centroids) |

Since QJL is removed, all 5 bits should go to MSE, giving 32 centroids and better reconstruction quality at a modest cost to compression ratio.

```suggestion
            bit_width: 5,
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: vortex-tensor/src/encodings/turboquant/array/scheme.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    fn scheme_name(&self) -> &'static str {
+        "vortex.tensor.turboquant"
+    }
+
+    fn matches(&self, canonical: &Canonical) -> bool {
+        let Canonical::Extension(ext) = canonical else {
+            return false;
+        };
+
+        TurboQuant::validate_dtype(ext.dtype()).is_ok()
```
Comment (lwwmanning) [inline vortex-tensor/src/encodings/turboquant/array/scheme.rs:49]: RFC 0033 Stage 1 specifies that the **scheme** minimum should be `dimension >= 128` for automatic selection, while the **array-level** minimum remains `dimension >= 3`:

> The threshold of 128 is conservative [...] The array-level minimum remains d=3 (for the Beta distribution to be well-defined), so users can still explicitly construct a TurboQuantArray at smaller dimensions. The scheme minimum (128) controls automatic selection only.

Currently `matches()` delegates to `validate_dtype()` which accepts `dimension >= 3`. This means the compressor will auto-select TurboQuant even for low-dimension vectors where SORF mixing quality degrades (only 18 butterfly stages at d=64), the per-vector norm overhead is proportionally large, and practical MSE may be worse.

Consider adding a separate dimension check here:

```rust
fn matches(&self, canonical: &Canonical) -> bool {
    let Canonical::Extension(ext) = canonical else {
        return false;
    };

    let Ok(ext_dtype) = TurboQuant::validate_dtype(ext.dtype()) else {
        return false;
    };

    // Scheme auto-selection only for dimension >= 128 (RFC 0033).
    // Smaller dimensions remain available via explicit TurboQuantArray construction.
    extension_list_size(ext_dtype).map_or(false, |d| d >= 128)
}
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: vortex-tensor/src/encodings/turboquant/vtable.rs [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    fn buffer(_array: ArrayView<Self>, idx: usize) -> BufferHandle {
+        vortex_panic!("TurboQuantArray buffer index {idx} out of bounds")
+    }
+
+    fn buffer_name(_array: ArrayView<Self>, _idx: usize) -> Option<String> {
+        None
+    }
+
+    fn serialize(array: ArrayView<'_, Self>) -> VortexResult<Option<Vec<u8>>> {
+        Ok(Some(vec![array.bit_width]))
```
Comment (lwwmanning) [inline vortex-tensor/src/encodings/turboquant/vtable.rs:176]: The metadata is serialized as a raw single byte. Every other encoding in this crate and across the Vortex codebase uses protobuf (`prost::Message`) for metadata serialization — including `FixedShapeTensor` in this same crate (see `fixed_shape/proto.rs`).

A bare byte has no framing or versioning. While adding `block_size`/`num_blocks` (RFC Stage 2 forward-compat fields) can be deferred, switching to protobuf *itself* should ideally happen now. A raw byte cannot be extended backward-compatibly — adding a second field later means existing serialized data (a single `0x04` byte) would need a migration path. Protobuf, on the other hand, handles unknown fields gracefully.

Suggested: use a prost message (even with just `bit_width` for now), consistent with the rest of the codebase. The Stage 2 fields can be added later as new proto fields with defaults.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: vortex-tensor/src/encodings/turboquant/compute/cosine_similarity.rs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+/// Convert an f32 value to `T`, returning `T::zero()` if the conversion fails.
+///
+/// This helper exists because `half::f16` has an inherent `from_f32` method that shadows
+/// the [`FromPrimitive`] trait method, causing compilation errors when used inside
+/// [`match_each_float_ptype!`].
+#[inline]
+fn f32_to_t<T: FromPrimitive + Zero>(v: f32) -> T {
+    // TODO(connor): Is this actually correct? How should we handle f64 overflow?
+    FromPrimitive::from_f32(v).unwrap_or_else(T::zero)
+}
```
Comment (lwwmanning) [inline vortex-tensor/src/encodings/turboquant/compute/cosine_similarity.rs]: This TODO should be resolved before merge. Analyzing the cases:
- `T = f64`: `FromPrimitive::from_f32` is lossless (f32 → f64 always succeeds), so no issue.
- `T = f32`: identity, no issue.
- `T = f16`: `FromPrimitive::from_f32` will return `None` for values outside f16 range (±65504), falling back to `T::zero()`. This silently drops the result. For cosine similarity (values in [-1, 1]) this is fine. For dot products scaled by norms, large norms could produce values outside f16 range.

Consider at minimum adding a comment documenting that the f16 case is safe for cosine similarity (bounded [-1, 1]) but could silently zero-out extreme dot product values. Or convert via `half::f16::from_f32()` which saturates to `±inf` rather than zeroing.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: vortex-tensor/src/encodings/turboquant/compute/cosine_similarity.rs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+//! Approximate cosine similarity in the quantized domain.
+//!
+//! Since the SRHT is orthogonal, inner products are preserved in the rotated
+//! domain. For two vectors from the same TurboQuant column (same rotation and
+//! centroids), we can compute the dot product of their quantized representations
+//! without full decompression:
+//!
+//! ```text
+//! cos_approx(a, b) = sum(centroids[code_a[j]] × centroids[code_b[j]])
+//! ```
```
Comment (lwwmanning) [inline vortex-tensor/src/encodings/turboquant/compute/cosine_similarity.rs]: Nit: This doc says "same rotation and centroids" but the implementation at lines 78-83 reads centroids from both arrays independently (correctly supporting different codebooks). The inner product preservation guarantee from the SRHT only holds if both arrays were rotated with the **same** rotation matrix, but this isn't validated.

If two TurboQuant arrays were encoded with different seeds (different SRHT rotations), the quantized inner product would be meaningless — you'd be dotting coordinates that were projected into different rotated spaces.

Consider either:
1. Validating that rotation signs match (or at least that dimensions match, which is done), and documenting the assumption that arrays from the same column share a rotation.
2. Updating the doc to clarify: "This function assumes both arrays share the same SRHT rotation (i.e., were encoded from the same TurboQuant column). Results are meaningless if the rotations differ."

---
_Generated by [Claude Code](https://claude.ai/code)_

File: vortex-tensor/src/encodings/turboquant/array/data.rs [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+    /// - `norms` is a primitive array whose ptype matches the element type of the Vector's storage
+    ///   dtype. The nullability must match `dtype.nullability()`. Norms carry the validity of the
+    ///   entire array, since null vectors have null norms.
+    /// - `centroids` is a non-nullable `PrimitiveArray<f32>` whose length is a power of 2 in
+    ///   `[2, 256]` (i.e., `2^bit_width` for bit_width 1-8), or empty for degenerate arrays.
+    /// - `rotation_signs` has `3 * padded_dim` elements, or is empty for degenerate arrays.
+    /// - For degenerate (empty) arrays: all children must be empty.
+    ///
+    /// Violating these invariants may produce incorrect results during decompression.
+    pub unsafe fn new_unchecked(
```
Comment (lwwmanning) [inline vortex-tensor/src/encodings/turboquant/array/data.rs:107]: Nit: This is `pub` but other encodings in the codebase (e.g., `BitPackedData`, `SparseData`) make their `new_unchecked` `pub(crate)`. Since this bypasses validation, restricting visibility reduces the risk of external callers constructing invalid arrays.

```suggestion
    pub(crate) unsafe fn new_unchecked(
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: vortex-tensor/src/encodings/turboquant/compress.rs [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+        PType::F16 => Ok(primitive
+            .as_slice::<half::f16>()
+            .iter()
+            .map(|&v| f32::from(v))
+            .collect()),
+        PType::F32 => Ok(primitive),
+        PType::F64 => Ok(primitive
+            .as_slice::<f64>()
+            .iter()
+            .map(|&v| v as f32)
```
Comment (lwwmanning) [inline vortex-tensor/src/encodings/turboquant/compress.rs:71]: Nit: The f64→f32 truncation here is intentional (quantization happens in f32), but for f64 inputs with values outside f32 range (~3.4e38), this silently produces `inf`/`-inf` which would then corrupt the normalization and quantization. In practice, neural network embeddings are nowhere near these ranges, but a brief comment documenting the assumption (or a checked cast) would help future readers.

Also: for f16→f32, `f32::from(v)` is lossless — good.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: vortex-tensor/src/encodings/turboquant/tests.rs
Comment (lwwmanning) [inline vortex-tensor/src/encodings/turboquant/tests.rs:1]: The test suite is thorough — good coverage of roundtrip correctness, MSE quality bounds, edge cases (empty, single-row, all-zero, nullable), compute pushdowns (slice, take, scalar_at), serde roundtrip, and L2 norm readthrough.

A few gaps to consider:

1. **No f16 input test.** The code handles f16→f32 upcast in `extract_f32_elements` and f16 norms in the decode path, but there's no test exercising this. An f16 roundtrip test would cover a real-world embedding scenario and catch any `half::f16` trait shadowing issues (like the one mentioned in `f32_to_t`).

2. **No test for `dot_product_quantized_column`.** Only cosine similarity is tested in the quantized domain. The dot product path multiplies by norms which exercises different code.

3. **No test for `TurboQuantScheme::compress()`.** The scheme integration (auto-selection through the compressor) isn't tested end-to-end.

4. **No large-dimension test.** The largest dimension tested is 256 (in MSE tests). Testing d=768 or d=1024 would validate the padding behavior for common embedding dimensions and stress the SRHT at higher butterfly counts.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: vortex-tensor/src/encodings/turboquant/array/centroids.rs [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+const INTEGRATION_POINTS: usize = 1000;
+
+/// Max-Lloyd convergence threshold.
+const CONVERGENCE_EPSILON: f64 = 1e-12;
+
+/// Maximum iterations for Max-Lloyd algorithm.
+const MAX_ITERATIONS: usize = 200;
+
+/// Global centroid cache keyed by (dimension, bit_width).
+static CENTROID_CACHE: LazyLock<DashMap<(u32, u8), Vec<f32>>> = LazyLock::new(DashMap::default);
```
Comment (lwwmanning) [inline vortex-tensor/src/encodings/turboquant/array/centroids.rs:30]: Minor: This is a process-global unbounded cache. While entries are tiny (max 256 f32s each) and the key space is bounded in practice (few distinct padded dimensions × 8 bit widths), it's worth documenting the intentional unbounded-ness — especially for library consumers who might wonder about memory leaks in long-running processes.

Also worth noting: the cache is keyed by `(dimension, bit_width)` where `dimension` is passed as `padded_dim` from `compress.rs:124`. So dim=768 and dim=900 both map to padded_dim=1024 and share cached centroids — correct behavior.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: vortex-tensor/src/encodings/turboquant/mod.rs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+//!
+//! TurboQuant minimizes mean-squared reconstruction error (1-8 bits per coordinate)
+//! using MSE-optimal scalar quantization with an SRHT rotation for coordinate independence.
+//!
+//! # Theoretical error bounds
+//!
+//! For unit-norm vectors quantized at `b` bits per coordinate, the paper's Theorem 1
+//! guarantees normalized MSE distortion:
+//!
+//! > `E[||x - x_hat||^2 / ||x||^2] <= (sqrt(3) * pi / 2) / 4^b`
```
Comment (lwwmanning) [inline vortex-tensor/src/encodings/turboquant/mod.rs:23]: Worth noting in the doc: the theoretical MSE bound from the paper (Theorem 1) is proved for **Haar-distributed random orthogonal matrices** (QR of Gaussian), not SORF/SRHT. The Vortex implementation uses SRHT as a practical O(d log d) approximation. The RFC acknowledges this gap explicitly:

> "Crucially, Theorem 1 is proved for true random orthogonal matrices (QR of Gaussian), not SORF."

The tests empirically validate that the bound holds with SRHT (which is reassuring), but mentioning this caveat in the doc would set correct expectations for users who read the paper.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by lwwmanning — CHANGES_REQUESTED** [review]
# Detailed Review — TurboQuant Encoding (RFC 0033 Stage 1)

This PR implements the TurboQuant lossy vector quantization algorithm ([arXiv:2504.19874](https://arxiv.org/abs/2504.19874)) as a new encoding in the `vortex-tensor` crate. It is intended to be Stage 1 of [RFC 0033](https://github.com/vortex-data/rfcs/blob/main/proposed/0033-block-turboquant.md). Overall this is well-engineered work — the algorithm is correctly implemented, the code follows Vortex conventions, tests are thorough, and the documentation is excellent. Requesting changes for a few RFC compliance issues.

## RFC 0033 Stage 1 Compliance

| RFC Requirement | Status | Details |
|---|---|---|
| QJL support removed | ✅ | No QJL code present |
| 4 slots (codes, norms, centroids, rotation_signs) | ✅ | Exactly 4 slots |
| Scheme default: 5-bit MSE-only | ❌ | Default is 4-bit (`compress.rs:44`) |
| Norms dtype: same-or-wider (f64→f64, f32/f16→f32) | ✅ | Correct |
| Scheme minimum: dimension ≥ 128 | ❌ | `TurboQuantScheme::matches()` accepts dimension ≥ 3 |
| Metadata: protobuf for forward compat | ⚠️ | Raw single byte, not protobuf — should be switched now to avoid a migration path later |

## Strengths

- **Clean architecture.** Module structure follows established Vortex patterns (vtable macro, data struct with `try_new`/`new_unchecked`, named slot enum, separate compress/decompress). Consistent with BitPacked, RLE, Sparse, etc.
- **Correct SRHT implementation.** The 3-round structured random Hadamard transform is correctly implemented with XOR-based branchless sign application (auto-vectorizes to `vpxor`/`veor`), iterative Walsh-Hadamard butterfly, and proper normalization factor `1/(n·√n)`. Forward/inverse symmetry is verified by tests.
- **Thorough validation.** `TurboQuantData::validate()` checks codes dtype, norms dtype matching, centroids power-of-2 constraint, rotation signs length, and degenerate/empty invariants. Debug assertions in `new_unchecked` add an extra safety net.
- **Smart compute pushdowns.** Slice/take operate on per-row children (codes, norms) and clone shared children (centroids, rotation_signs). Quantized cosine similarity and dot product avoid full decompression. L2 norm readthrough is O(1) from stored norms.
- **Excellent test coverage** (911 lines): roundtrip, MSE quality bounds, edge cases, nullable vectors, serde roundtrip, compute pushdowns, L2 norm readthrough.
- **Good documentation.** Module docs include theoretical MSE bounds, compression ratio tables, and a working example. Per-function docs explain algorithmic context.

## Key Issues (see inline comments)

1. **Default bit_width should be 5, not 4** — RFC specifies "5-bit MSE-only (32 centroids)" 
2. **Scheme minimum should be dimension ≥ 128** — RFC specifies auto-selection only for d ≥ 128
3. **Metadata serialization should use protobuf** — raw byte can't be extended backward-compatibly; `FixedShapeTensor` in the same crate uses prost
4. **Unresolved TODO in `f32_to_t`** — should be resolved or documented before merge
5. **Cosine similarity doc claims "same rotation"** — but doesn't validate this; should clarify assumptions

## Algorithmic Note

The theoretical MSE bound (Theorem 1 in the paper) is proved for Haar-distributed random orthogonal matrices, not SORF/SRHT. The SRHT is a practical approximation. The RFC explicitly acknowledges this. The tests empirically validate the bound holds with SRHT, which is good — but worth noting in the module docs.

## Minor Items

- `new_unchecked` is `pub` — other encodings use `pub(crate)` 
- No f16 input roundtrip test
- No quantized dot product test (only cosine similarity)
- No `TurboQuantScheme::compress()` integration test
- Global centroid cache is unbounded (fine in practice, worth documenting)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: vortex-tensor/src/encodings/turboquant/compress.rs (hunk 1 above)
Comment (lwwmanning) [inline vortex-tensor/src/encodings/turboquant/compress.rs]: Follow-up to my earlier comment: after further discussion, the default bit_width should be **8** (near-lossless, MSE ~4e-5 per the table in the module docs). This diverges from the RFC's "5-bit" recommendation, but 8-bit is a safer default for a general-purpose encoding — users who want aggressive compression can explicitly opt into lower bit widths.

At 8 bits the compression ratio is still meaningful (3-4x for f32 inputs at common dimensions), and the reconstruction quality is near-lossless, making it appropriate as a default that "just works" without surprises.

```suggestion
            bit_width: 8,
```

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by lwwmanning — COMMENTED** [review]
Follow-up: after further consideration, the default `bit_width` should be **8** (near-lossless) rather than the RFC's 5. At 8 bits the normalized MSE is ~4e-5 — effectively transparent — while still achieving 3-4x compression on f32 data. This is a safer default for a general-purpose encoding; users who want more aggressive compression can explicitly configure lower bit widths.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by lwwmanning — COMMENTED** [review]
Claude diffed against RFC 33 and had a few minor comments that look valid/easy to fix before merging

File: vortex-tensor/src/encodings/turboquant/vtable.rs (hunk 3 above)
Comment (connortsui20) [reply vortex-tensor/src/encodings/turboquant/vtable.rs:176]: We don't care about this right now (dont need to be backwards compat yet), and really we might not actually need any metadata in the future with the upcoming vtable plans

File: vortex-tensor/src/encodings/turboquant/array/centroids.rs (hunk 8 above)
Comment (connortsui20) [reply vortex-tensor/src/encodings/turboquant/array/centroids.rs:30]: we can review this implementation later

**Comment by connortsui20** [comment]
So based on my reading of the review comments, the things we _actually_ want to change are:

- Default bitwidth changed to 8 (makes sense)
- A few more tests for f16, f64, very large dimensions, and test helper functions themselves
- Some docs need to be updated
- Dimension must be >= 128 (is there a theoretical backing for this number)?
- Fix the f32 cast semantics (we have some safety guarantees that need to be documented)

Everything else is either wrong or something we can think about later.

**Comment by connortsui20** [comment]
Comments addressed.

There is a small issue related to https://github.com/vortex-data/vortex/pull/7269#discussion_r3035622938

This is what my claude instance told me after seeing pretty high errors:

<details>

The theoretical guarantee. Theorem 1's MSE bound ((sqrt(3) * pi / 2) / 4^b) is proved for Haar-distributed random orthogonal matrices — matrices drawn uniformly from the orthogonal group O(d). The proof depends on two properties that Haar matrices give
   you exactly and SRHT gives you approximately:

  1. Uniform distribution on the sphere. For a Haar matrix Π and any unit vector x, Πx is exactly uniformly distributed on S^{d-1}. This means each coordinate of the rotated vector follows exactly the Beta((d-1)/2, (d-1)/2) distribution on [-1, 1]. Our
  Max-Lloyd centroids are computed for this distribution, so they're exactly optimal.

  With SRHT, the rotated coordinates are approximately Beta-distributed. The centroids are slightly suboptimal for the actual marginals.

  2. Near-independence of coordinates. The proof relies on coordinates of Πx being "nearly independent" in high dimensions (a deep result from high-dimensional probability, reference [55] in the paper — Vershynin). This ensures quantization errors across
   coordinates don't conspire to accumulate worse than expected.

  With SRHT, coordinates are uncorrelated but have different higher-order dependence structure. The 3-round structure (H·D₃·H·D₂·H·D₁) gives good mixing but it's not equivalent to full randomness.

  What we gain in exchange: O(d log d) computation and O(d) storage instead of O(d²) for both. The SRHT butterfly + XOR sign application auto-vectorizes into SIMD. For d=1024, that's a ~1000x speedup over a full matrix multiply.

  In practice the gap is small. Our tests show:
  - At d=128, 4-bit: MSE is well within the 1x theoretical bound
  - At d=1024, 4-bit: MSE exceeds the bound by ~20% (0.0127 vs 0.0106)
  - Mixing quality improves with dimension (more butterfly stages: 3×log₂(d))

  The RFC's fallback plan if SRHT ever proves insufficient is to use a full B×B random orthogonal matrix per block at Stage 2 block sizes (e.g., B=256 → 256KB storage per block)

</details>

We likely want to implement both (doesn't seem to be too hard) and then compare them later.

**Review by lwwmanning — APPROVED** [review]
